trace/trace.go

```diff
@@ -0,0 +1,215 @@
+// Package trace implement structured logging of requests
+package trace
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/mailgun/oxy/utils"
+)
+
+// Option is a functional option setter for Tracer
+type Option func(*Tracer) error
+
+// ErrorHandler is a functional argument that sets error handler of the server
+func ErrorHandler(h utils.ErrorHandler) Option {
+	return func(t *Tracer) error {
+		t.errHandler = h
+		return nil
+	}
+}
+
+// RequestHeaders adds request headers to capture
+func RequestHeaders(headers ...string) Option {
+	return func(t *Tracer) error {
+		t.reqHeaders = append(t.reqHeaders, headers...)
+		return nil
+	}
+}
+
+// ResponseHeaders adds response headers to capture
+func ResponseHeaders(headers ...string) Option {
+	return func(t *Tracer) error {
+		t.respHeaders = append(t.reqHeaders, headers...)
+		return nil
+	}
+}
+
+// Logger sets optional logger for trace used to report errors
+func Logger(l utils.Logger) Option {
+	return func(t *Tracer) error {
+		t.log = l
+		return nil
+	}
+}
+
+// Tracer records request and response emitting JSON structured data to the output
+type Tracer struct {
+	errHandler  utils.ErrorHandler
+	next        http.Handler
+	reqHeaders  []string
+	respHeaders []string
+	writer      io.Writer
+	log         utils.Logger
+}
+
+// New creates a new Tracer middleware that emits all the request/response information in structured format
+// to writer and passes the request to the next handler. It can optionally capture request and response headers,
+// see RequestHeaders and ResponseHeaders options for details.
+func New(next http.Handler, writer io.Writer, opts ...Option) (*Tracer, error) {
+	t := &Tracer{
+		writer: writer,
+		next:   next,
+	}
+	for _, o := range opts {
+		if err := o(t); err != nil {
+			return nil, err
+		}
+	}
+	if t.errHandler == nil {
+		t.errHandler = utils.DefaultHandler
+	}
+	if t.log == nil {
+		t.log = utils.NullLogger
+	}
+	return t, nil
+}
+
+func (t *Tracer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	start := time.Now()
+	pw := &utils.ProxyWriter{W: w}
+	t.next.ServeHTTP(pw, req)
+
+	l := t.newRecord(req, pw, time.Since(start))
+	bytes, err := json.Marshal(l)
+	if err != nil {
+		t.log.Errorf("Failed to marshal request: %v", err)
+		return
+	}
+	t.writer.Write(bytes)
```

Comment:
There could also be an error here. Maybe we could log that, too? Don't feel strongly either way.

Using `json.NewEncoding(t.writer).Encode(l)` would give you a single error value to check instead of two, and still catches both Marshal errors and Write errors.